File: apps/api/api/routes/workspaces.py
"""Workspace (group chat) API routes."""

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from integrations import require_user_context
from workspaces import (
    create_workspace,
    list_workspace_members,
    list_workspace_messages,
    list_workspaces,
    resync_workspace_context,
    send_workspace_message,
)

router = APIRouter(prefix="/workspaces", tags=["workspaces"])


class CreateWorkspaceInput(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    member_user_ids: list[str] = Field(default_factory=list)
    purpose: str | None = None
    context_md: str | None = None
    loombot_mode: str | None = None


class SendWorkspaceMessageInput(BaseModel):
    message: str = Field(min_length=1)


@router.get("")
async def workspaces(
    ctx: tuple[str, str] = Depends(require_user_context),
) -> list[dict]:
    return await list_workspaces(*ctx)


@router.post("")
async def create(
    request: CreateWorkspaceInput,
    ctx: tuple[str, str] = Depends(require_user_context),
) -> dict:
    return await create_workspace(
        org_id=ctx[0],
        user_id=ctx[1],
        name=request.name,
        member_user_ids=request.member_user_ids,
        purpose=request.purpose,
        context_md=request.context_md,
        loombot_mode=request.loombot_mode,
    )


@router.post("/{workspace_id}/resync-context")
async def resync_context(
    workspace_id: str,
    ctx: tuple[str, str] = Depends(require_user_context),
) -> dict:
    return await resync_workspace_context(
        org_id=ctx[0],
        user_id=ctx[1],
        workspace_id=workspace_id,
    )


@router.get("/{workspace_id}/members")
async def members(
    workspace_id: str,
    ctx: tuple[str, str] = Depends(require_user_context),
) -> list[dict]:
    return await list_workspace_members(*ctx, workspace_id)


@router.get("/{workspace_id}/messages")
async def messages(
    workspace_id: str,
    ctx: tuple[str, str] = Depends(require_user_context),
) -> list[dict]:
    return await list_workspace_messages(*ctx, workspace_id)


@router.post("/{workspace_id}/messages")
async def send_message(
    workspace_id: str,
    request: SendWorkspaceMessageInput,
    ctx: tuple[str, str] = Depends(require_user_context),
) -> dict:
    return await send_workspace_message(
        org_id=ctx[0],
        user_id=ctx[1],
        workspace_id=workspace_id,
        body=request.message,
    )

File: apps/web/src/services/workspaces.ts
import { apiFetch } from "@/lib/api";

export interface Workspace {
  workspace_id: string;
  name: string;
  kind: "org_wide" | "group" | string;
  created_by?: string | null;
  created_at?: string;
  updated_at?: string;
  member_count: number;
  last_message?: string | null;
  last_message_at?: string | null;
  last_sender_name?: string | null;
  purpose?: string | null;
  context_md?: string | null;
  context_synced_at?: string | null;
  loombot_mode?: "context_only" | "org_knowledge" | string;
}

export interface WorkspaceMessage {
  message_id: string;
  sender_user_id?: string | null;
  sender_type: "user" | "bot" | string;
  sender_name: string;
  body: string;
  created_at: string;
}

export interface WorkspaceMember {
  user_id: string;
  name: string;
  email: string;
  role: string;
}

export interface CreateWorkspaceInput {
  name: string;
  member_user_ids?: string[];
  purpose?: string | null;
  context_md?: string | null;
  loombot_mode?: "context_only" | "org_knowledge" | string | null;
}

export async function listWorkspaces(): Promise<Workspace[]> {
  const response = await apiFetch("/workspaces");
  if (!response.ok) throw new Error("Could not load workspaces.");
  return response.json();
}

export async function createWorkspace(
  nameOrInput: string | CreateWorkspaceInput,
  memberUserIds: string[] = [],
): Promise<Workspace> {
  const body: CreateWorkspaceInput =
    typeof nameOrInput === "string"
      ? { name: nameOrInput, member_user_ids: memberUserIds }
      : {
          name: nameOrInput.name,
          member_user_ids: nameOrInput.member_user_ids ?? [],
          purpose: nameOrInput.purpose,
          context_md: nameOrInput.context_md,
          loombot_mode: nameOrInput.loombot_mode,
        };
  const response = await apiFetch("/workspaces", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    let detail = "Could not create the workspace.";
    try {
      const payload = await response.json();
      detail = (payload?.detail as string) || detail;
    } catch {
      /* keep default */
    }
    throw new Error(detail);
  }
  return response.json();
}

export async function resyncWorkspaceContext(
  workspaceId: string,
): Promise<Workspace> {
  const response = await apiFetch(`/workspaces/${workspaceId}/resync-context`, {
    method: "POST",
  });
  if (!response.ok) {
    let detail = "Could not resync CONTEXT.md.";
    try {
      const payload = await response.json();
      detail = (payload?.detail as string) || detail;
    } catch {
      /* keep default */
    }
    throw new Error(detail);
  }
  return response.json();
}

export async function listWorkspaceMessages(
  workspaceId: string,
): Promise<WorkspaceMessage[]> {
  const response = await apiFetch(`/workspaces/${workspaceId}/messages`);
  if (!response.ok) throw new Error("Could not load workspace messages.");
  return response.json();
}

export async function listWorkspaceMembers(
  workspaceId: string,
): Promise<WorkspaceMember[]> {
  const response = await apiFetch(`/workspaces/${workspaceId}/members`);
  if (!response.ok) throw new Error("Could not load workspace members.");
  return response.json();
}

export async function sendWorkspaceMessage(
  workspaceId: string,
  message: string,
): Promise<{ message: WorkspaceMessage; bot_message?: WorkspaceMessage | null }> {
  const response = await apiFetch(`/workspaces/${workspaceId}/messages`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ message }),
  });
  if (!response.ok) throw new Error("Could not send the message.");
  return response.json();
}
